src/meepgeom.cpp

```diff
@@ -2853,188 +2853,180 @@ void material_grids_addgradient_point(double *v, vector3 p, double scalegrad, ge
-                     md->medium_2.epsilon_offdiag.y.re != 0 ||
-                     md->medium_2.epsilon_offdiag.z.re != 0) {
-            /* we need to restrict the adjoint fields to
-            the two nodes of interest (which requires a factor
-            of 0.5 to scale), and interpolate the forward fields
-            to the same two nodes (which requires another factor of 0.5).
-            Then we perform our inner product at these nodes.
-            */
-            std::complex<double> fwd_avg, fwd1, fwd2;
-            ptrdiff_t fwd1_idx, fwd2_idx;
-
-            //identify the first corner of the forward fields
-            meep::ivec fwd_p = ip + gv.iyee_shift(forward_c) - gv.iyee_shift(adjoint_c);
-
-            //identify the other three corners
-            meep::ivec unit_a  = unit_ivec(gv.dim,component_direction(adjoint_c));
-            meep::ivec unit_f  = unit_ivec(gv.dim,component_direction(forward_c));
-            meep::ivec fwd_pa  = (fwd_p + unit_a*2);
-            meep::ivec fwd_pf  = (fwd_p - unit_f*2);
-            meep::ivec fwd_paf = (fwd_p + unit_a*2 - unit_f*2);
-
-            //identify the two eps points
-            meep::ivec ieps1 = (fwd_p + fwd_pf) / 2;
-            meep::ivec ieps2 = (fwd_pa + fwd_paf) / 2;
-
-            //operate on the first eps node
-            fwd1_idx = get_idx_from_ivec(gv.dim, start_ivec, the_stride, fwd_p);
-            fwd1 = 0.5 * meep::cdouble(GET_FIELDS(fields_f, ci_forward, f_i, fwd1_idx));
-            fwd2_idx = get_idx_from_ivec(gv.dim, start_ivec, the_stride, fwd_pf);
-            fwd2 = 0.5 * meep::cdouble(GET_FIELDS(fields_f, ci_forward, f_i, fwd2_idx));
-            fwd_avg = fwd1 + fwd2;
-            meep::vec eps1 = gv[ieps1];
-            cyl_scale = (gv.dim == meep::Dcyl) ? eps1.r() : 1;
-            material_grids_addgradient_point(
-                v+ng*f_i, vec_to_vector3(eps1), scalegrad*cyl_scale, geps,
-                adjoint_c, forward_c, fwd_avg, 0.5*adj, frequencies[f_i], gv, du);
-
-            //operate on the second eps node
-            fwd1_idx = get_idx_from_ivec(gv.dim, start_ivec, the_stride, fwd_pa);
-            fwd1 = 0.5 * meep::cdouble(GET_FIELDS(fields_f, ci_forward, f_i, fwd1_idx));
-            fwd2_idx = get_idx_from_ivec(gv.dim, start_ivec, the_stride, fwd_paf);
-            fwd2 = 0.5 * meep::cdouble(GET_FIELDS(fields_f, ci_forward, f_i, fwd2_idx));
-            fwd_avg = fwd1 + fwd2;
-            meep::vec eps2 = gv[ieps2];
-            cyl_scale = (gv.dim == meep::Dcyl) ? eps2.r() : 1;
-            material_grids_addgradient_point(
-                v+ng*f_i, vec_to_vector3(eps2), scalegrad*cyl_scale, geps,
-                adjoint_c, forward_c, fwd_avg, 0.5*adj, frequencies[f_i], gv, du);
+            /* more complicated case requires interpolation/restriction */
+            } else if (
+                      md->do_averaging ||                         /* account for subpixel smoothing     */
```

Comment:
Let's try to use consistent formatting: no new line for the first clause of the `if` statement and also subsequent lines should be aligned properly.

Reply:
Fixed.